### PR TITLE
test(client): add "chunking-query" test suite, differentiating Driver Adapters from plain Rust drivers

### PR DESCRIPTION
**Pull request by jkomyno**
This PR closes https://github.com/prisma/team-orm/issues/394 and https://github.com/prisma/team-orm/issues/519.

In particular, we added a new functional test suite `chunking-query` that subsumes the test suites for issues 8832 and 9326, and adds `cockroachdb` to the test matrix.

Moreover, we now explicitly test scenarios in which Rust drivers get a `too many bind variables in prepared statement` database assertion error, but Driver Adapters don't.
These scenarios are explained by the new issues https://github.com/prisma/prisma/issues/21802 and https://github.com/prisma/prisma/issues/21803.

---

Note: the newly added tests are not applicable to PlanetScale, as Vitess supports at most 10k rows returned in a single query, and the test tries to retrieve 65k+ rows.

### CODE REVIEW

File: packages/client/tests/functional/too-many-bind-variables/tests.ts
Comment (janpio) [inline packages/client/tests/functional/too-many-bind-variables/tests.ts:64]: Shouldn't these work as the engine should take care of chunking to split the queries into multiple?

Comment (Jolg42) [reply packages/client/tests/functional/too-many-bind-variables/tests.ts:64]: That was also my expectation, same for this user in https://github.com/prisma/prisma/issues/21648#issuecomment-1783838900
Because we fixed something that looks similar in https://github.com/prisma/prisma/issues/8832#issuecomment-1260533913

Comment (miguelff) [reply packages/client/tests/functional/too-many-bind-variables/tests.ts:64]: This is indeed a bug, the Postgres extended protocol uses a short to identify the number of query parameters, but we know that at the engine level, because we know wether rust drivers will run. We could fix that in this PR, but I don't think it's in the scope of this issue, but a correctness one, as it's not a bug in driver adapters code.

Comment (miguelff) [reply packages/client/tests/functional/too-many-bind-variables/tests.ts:64]: Actually the investigation subject of https://github.com/prisma/team-orm/issues/394 is done without the need to merge this PR and the outcome should be a correctness issue to fix the binding problem when using rust drivers.

Comment (jkomyno) [reply packages/client/tests/functional/too-many-bind-variables/tests.ts:64]: We have had previous indications of Driver Adapters not using the query chunking logic defined in the Prisma engines.
Keeping [this](https://github.com/prisma/team-orm/issues/394#issuecomment-1787218882) in mind, I've answered in a top-level comment for future visibility. See: https://github.com/prisma/prisma/pull/21679#issuecomment-1787315586.

File: packages/client/tests/functional/too-many-bind-variables/tests.ts
Comment (jkomyno) [inline packages/client/tests/functional/too-many-bind-variables/tests.ts:77]: @miguelff, this shows Driver Adapters failing on queries with `65536` parameters

File: packages/client/tests/functional/too-many-bind-variables/tests.ts
Comment (jkomyno) [inline packages/client/tests/functional/too-many-bind-variables/tests.ts:60]: @miguelff, this shows Driver Adapters succeeding on queries with `32768` parameters, despite Rust Drivers failing on the same number

**Comment by jkomyno** [comment]
Note: currently investigating some divergence between local execution and CI runs.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [inline packages/client/tests/functional/chunking-query/tests.ts:72]: I think these are all specific to the first `describe`, could we move them there?

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [inline packages/client/tests/functional/chunking-query/tests.ts:242]: These are the failing cases, can we wrap them into their own `describe`?
(That would also allow us to move `selectWith2InFilters` into here as it is used in both `describeIf` that will then be children of it)

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [inline packages/client/tests/functional/chunking-query/tests.ts:28]: See https://github.com/prisma/prisma/pull/21679/files#r1383648125

File: packages/client/tests/functional/issues/8832/tests.ts
Comment (janpio) [inline packages/client/tests/functional/issues/8832/tests.ts:157]: This is the important bit that is missing in the new test file, right?

File: packages/client/tests/functional/issues/9326/tests.ts
Comment (janpio) [inline packages/client/tests/functional/issues/9326/tests.ts:57]: This is the important bit that is missing in the new test file, right?

File: packages/client/tests/functional/issues/8832/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/issues/8832/tests.ts:157]: This troublesome bit has been reworked so that tests are run on `JS_NEON` / `JS_PG` as well as Rust drivers in the new `chunking-queries` test suite.

File: packages/client/tests/functional/issues/9326/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/issues/9326/tests.ts:57]: This troublesome bit has been reworked so that tests are run on `JS_NEON` / `JS_PG` as well as Rust drivers in the new `chunking-queries` test suite.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:242]: > These are the failing cases

Not all of them are failing: "With Driver Adapters only > Selecting 32768 ids at once in two inclusive disjunct filters works" is succeeding. Where should that be? How many `describe` blocks would you recommend having here?

Comment (janpio) [reply packages/client/tests/functional/chunking-query/tests.ts:242]: Enough to structure the code in a way that one can understand what the "helper functions" are used for.

Generally these are the queries where chunking currently does not work though, and for driver adapters it only works as the driver can handle more than what our chunking would create (as our chunking is based on the tokio-postgres capabilities, which are artificially low). So a grouping describe "chunking logic does not trigger" would work for me here.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [inline packages/client/tests/functional/chunking-query/tests.ts:260]: What does "unreachable assertions" mean here?
What happens for Data Proxy / Mini Proxy?

Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:260]: As written elsewhere, the CI fails on *Proxy, as the test execution runs "unreachable assertions" (which appear in the code as `expect(true).toBe(false)`). The `skipDataProxy` settings avoids this problem.

I'm very confused about why such gimmick is needed in the first place, though.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts]: (I have implemented this refactoring)

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:262]: I've found the problem, and I'm fixing it.

Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts]: It's probably fixed now. I've discovered that, if we run the MySQL-equivalent snippet of https://github.com/prisma/prisma/issues/21802 (`n = 32767`), we don't get an error at all (as opposed to Postgres/CockroachDB, where we get the error "too many bind variables in prepared statement"). However, if we increase `n` by 1 (`n = 32767`), we get the error "Prepared statement contains too many placeholders", which is analogous to Postgres' "too many bind variables in prepared statement"

Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts]: Now I'm investigating a potential race condition: sometimes (with Neon), we get an error message containing "bind message has 5 parameter formats but 0 parameters", although the expectation is "bind message has 32767 parameter formats but 0 parameters".

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [inline packages/client/tests/functional/chunking-query/tests.ts:17]: Why does it takes ages with Neon?
What is the `(while `js_pg` is much)` supposed to mean? Don't understand.

Comment (janpio) [reply packages/client/tests/functional/chunking-query/tests.ts:17]: Where are we actually exclusing `js_neon`? Just outdated comment?

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [inline packages/client/tests/functional/chunking-query/tests.ts:16]: This should probably be the explanation in the `optOut` at the bottom instead instead of a comment here?

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [inline packages/client/tests/functional/chunking-query/tests.ts:160]: Unrelated to this PR:
1. All these tests would benefit from an explanation _why_ they succeed. Either because the number of params is fine for one query, or because we detect the too high number and split into multiple queries, right?
2. Additionally it would be really nice if we would really confirm the number of queries that are executed. That would also "document" what the chunking is doing.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (janpio) [reply packages/client/tests/functional/chunking-query/tests.ts]: Including MysQL definitely makes this more messy, but good learning! Seems the missing chunking behaves different there than with PostgreSQL. Good call 👍

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:16]: Yes, the placement of this comment has historical reasons that no longer apply. I'm moving it below.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:17]: The comment for `js_neon` no longer applies. I had added the comment in a previous iteration, where my environment loaded Node.js rather than Node.js 20. This resulted in WebSockets silently failing to connect to the database, resulting in a stalled CI (until the 180s time limit of our test suite is reached).
I'll remove the comment, thanks for spotting it out.

Note: I've double-checked that the CI for Driver Adapters uses Node.js 20 only.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:157]: https://github.com/prisma/prisma/issues/21827

**Review by janpio — APPROVED** [review]
Look good to me now. 
Wait for actual Eng review though before merge, please.

**Review by miguelff — APPROVED** [review]

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (Jolg42) [inline packages/client/tests/functional/chunking-query/tests.ts:66]: Nit: I'm more used to write a transaction for this, because the order is guaranteed and it makes the code easier I think like
```suggestion
        await prisma.$transaction([
          prisma.tagsOnPosts.deleteMany(), 
          prisma.post.deleteMany(), 
          prisma.tag.deleteMany()
        ])
```

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (Jolg42) [inline packages/client/tests/functional/chunking-query/tests.ts:227]: It's missing an expect like 
```suggestion
        test('Selecting 32768 ids at once in two inclusive disjunct filters works', async () => {
          const ids = generatedIds(32768)
          await expect(selectWith2InFilters(ids)).resolves.toBeTruthy()
        })
```
`toBeTruthy()` is an example, maybe there is a better thing to expect here but I still need to read through things

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (Jolg42) [inline packages/client/tests/functional/chunking-query/tests.ts:244]:
```suggestion
        test('Selecting 65536 ids at once in two inclusive disjunct filters results in error', async () => {
          expect.assertions(1)
          const ids = generatedIds(65536)

          try {
            // this should throw 
            await selectWith2InFilters(ids)
          } catch (error) {
            const e = error as Error

            if (['postgresql', 'cockroachdb'].includes(provider)) {
              expect(e.message).toContain('bind message has 32767 parameter formats but 0 parameters')
            }
          }
        })
```
When adding expects inside a `try catch` we can add an expectation for the number of assertions `expect.assertions(1)`
Also, it's the pattern we follow in other tests so for clarity and consistency it's better I think.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:227]: If that rejected, an error would be thrown, and the test would fail (so we'd see that in the CI).
Note: this is a straight-up copy of previously existing tests.

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:244]: While `expect.assertions(_)` is used elsewhere for the reasons you suggest, in this case, it would add more noise than benefits. Why? Because the `expect()` in the `catch` branch is only called conditionally, so the number of expected assertions should also be computed conditionally. Thanks for the comment, though

File: packages/client/tests/functional/chunking-query/tests.ts
Comment (jkomyno) [reply packages/client/tests/functional/chunking-query/tests.ts:252]: In the original tests, only `postgres` was tested. Now, we test `cockroachdb` (as it uses the same Rust driver) and `mysql` (added originally to test `planetscale`, which is however skipped atm).

There was no specific reason to very `sqlserver` as well, as that's out of scope for the issue this PR solves, but that could change in the future.

**Pull request merged**
